File: frontend/src/types/session.ts
export interface GuidingQuestionResponse {
  guiding_question_id: string;
  text_response: string | null;
  audio_file: string | null;
  transcript: string | null;
  transcript_confidence: number | null;
}

export interface FRQRunAttempt {
  attempt_number: number;
  code_snapshot: string;
  stdout: string;
  stderr: string;
  compile_errors: string[];
  passed: boolean;
  timestamp: string;
}

export interface FRQPartResponse {
  part: string;
  final_code: string;
  passed: boolean;
  total_runs: number;
  total_time_seconds: number;
  time_to_first_keystroke_seconds: number;
  run_attempts: FRQRunAttempt[];
  compile_errors_encountered: string[];
  line_time_map: Record<string, number>;
  line_edit_counts: Record<string, number>;
  struggled_lines: number[];
}

export interface QuestionResponse {
  question_id: string;
  question_type: 'multiple_choice' | 'code_trace' | 'frq';
  selected_answer: string | null;
  frq_parts: FRQPartResponse[];
  is_correct: boolean;
  attempt_number: number;      // 1 = first try, 2 = second try
  score_weight: number;        // 1.0 | 0.7 | 0.0
  time_spent_seconds: number;
  guiding_question_responses: GuidingQuestionResponse[];
}

export interface SessionSaveRequest {
  session_id: string;
  student_name: string;
  test_id: string;
  iteration: number;
  started_at: string;
  completed_at: string;
  responses: QuestionResponse[];
}

export interface SessionMetadata {
  session_id: string;
  student_name: string;
  test_id: string;
  iteration: number;
  phase: 'assessment' | 'report_ready' | 'practice_ready' | 'quiz_ready' | 'complete';
  timestamp: string;
  total_questions: number;
  completed: boolean;
  study_completed: boolean;
}

File: backend/models/session.py
from pydantic import BaseModel
from typing import Optional, Literal

# ── Iteration phase ───────────────────────────────────────────────────────────
IterationPhase = Literal[
    "assessment",     # taking the test
    "report_ready",   # test complete, report generated
    "practice_ready", # practice questions generated (Phase 3)
    "quiz_ready",     # practice quiz generated (Phase 3)
    "complete",       # full iteration done
]

# ── Session start ─────────────────────────────────────────────────────────────
class SessionStartRequest(BaseModel):
    student_name: str
    test_id: str
    total_questions: int

# ── Guiding question response (MC questions) ──────────────────────────────────
class GuidingQuestionResponse(BaseModel):
    guiding_question_id: str
    text_response: Optional[str] = None
    audio_file: Optional[str] = None
    transcript: Optional[str] = None
    transcript_confidence: Optional[float] = None

# ── FRQ run attempt ───────────────────────────────────────────────────────────
class FRQRunAttempt(BaseModel):
    attempt_number: int
    code_snapshot: str
    stdout: str
    stderr: str
    compile_errors: list[str] = []
    passed: bool
    timestamp: str

# ── FRQ part response (rich behavioral metadata) ──────────────────────────────
class FRQPartResponse(BaseModel):
    part: str                                   # "a", "b", "c"
    final_code: str
    passed: bool
    total_runs: int
    total_time_seconds: int
    time_to_first_keystroke_seconds: int
    run_attempts: list[FRQRunAttempt] = []
    compile_errors_encountered: list[str] = []  # all unique errors seen
    line_time_map: dict[str, int] = {}           # line number → seconds
    line_edit_counts: dict[str, int] = {}        # line number → edit count
    struggled_lines: list[int] = []              # lines with high time or edit count

# ── MC question response ──────────────────────────────────────────────────────
class QuestionResponse(BaseModel):
    question_id: str
    question_type: Literal["multiple_choice", "code_trace", "frq"] = "multiple_choice"
    selected_answer: Optional[str] = None        # MC / code_trace
    frq_parts: list[FRQPartResponse] = []        # FRQ only
    is_correct: bool
    time_spent_seconds: int
    guiding_question_responses: list[GuidingQuestionResponse] = []

# ── Full session save payload ─────────────────────────────────────────────────
class SessionSaveRequest(BaseModel):
    session_id: str
    student_name: str
    test_id: str
    iteration: int
    started_at: str
    completed_at: str
    responses: list[QuestionResponse]
